{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Read the Reference Genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AGCTTTTCATTCTGACTGCAACGGGCAATATGTCTCTGTGTGGATTAAAAAAAGAGTGTCTGATAGCAGCTTCTGAACTG\n",
      "GTTACCTGCCGTGAGTAAATTAAAATTTTATTGACTTAGGTCACTAAATACTTTAACCAATATAGGCATAGCGCACAGAC\n",
      "AGATAAAAATTACAGAGTACACAACATCCATGAAACGC\n"
     ]
    }
   ],
   "source": [
    "from IPython.display import clear_output\n",
    "import math\n",
    "import numpy as np\n",
    "nrow = 64\n",
    "ncol = 64\n",
    "\n",
    "fname = \"GCF_000005845.2_ASM584v2_genomic.fna\"\n",
    "\n",
    "f = open(fname,'r')\n",
    "lines = f.readlines()[1:]\n",
    "f.close()\n",
    "\n",
    "genome = ''.join(lines)\n",
    "print(genome[0:200])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Convert the genome code to bits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "g = 'G'\n",
    "gbitlen = 3\n",
    "\n",
    "# Converts each genome base into bits. The choice of this encoding is based on three different principles:\n",
    "# 1. Since we are using TCAM we want an equal hamming distance with each substitution\n",
    "# 2. Since transitions occur (2.1x) more often than transversions. \n",
    "#    Transversions will have 2x more penalty than transitions.\n",
    "#    Transitions have a hamming distance of 1 and transversions have a hamming distance of 2.\n",
    "# 3. Genomes are double stranded the reads can come from any of the two strands. \n",
    "#    The other strand will be a reverse complement of the other strand.\n",
    "#    The encoding helps in doing this, in such a way that the reverse of the encoding is already its complement.\n",
    "def base2bits(g):\n",
    "    code = '' #'****'\n",
    "    #print(g)\n",
    "    if g == 'A':\n",
    "        code = '0*01'\n",
    "    elif g == 'C':\n",
    "        code = '11*0'\n",
    "    elif g == 'T':\n",
    "        code = '10*0'\n",
    "    elif g == 'G':\n",
    "        code = '0*11'\n",
    "    return code    \n",
    "\n",
    "# Converts a genome string into a genome bit string\n",
    "def genome2bits(genome):\n",
    "    gbits = '' #'x' * (len(genome) * gbitlen)\n",
    "    for g in genome: #range(len(genome):\n",
    "        gbits = gbits + base2bits(g)\n",
    "    return gbits    \n",
    "    \n",
    "gbits = genome2bits(genome)        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AGCTTTTCAT\n"
     ]
    }
   ],
   "source": [
    "print(genome[0:10])\n",
    "\n",
    "def base2mask(g):\n",
    "    code = '' #'****'\n",
    "    mask = ''\n",
    "    #print(g)\n",
    "    if g == 'A':\n",
    "        code = '0001'\n",
    "        mask = '0100'\n",
    "    elif g == 'C':\n",
    "        code = '1100'\n",
    "        mask = '0010'\n",
    "    elif g == 'T':\n",
    "        code = '1000'\n",
    "        mask = '0010'\n",
    "    elif g == 'G':\n",
    "        code = '0011'\n",
    "        mask = '0100'\n",
    "    elif g == 'N':\n",
    "        code = '0000'\n",
    "        mask = '1111'\n",
    "    return code, mask\n",
    "\n",
    "def genome2mask(genome):\n",
    "    code = '' #'x' * (len(genome) * gbitlen)\n",
    "    mask = ''\n",
    "    codeint = []\n",
    "    maskint = []\n",
    "        \n",
    "    if len(genome)%16 != 0:\n",
    "        genome = genome + 'N'*(16-len(genome)%16)\n",
    "    \n",
    "    for g in genome: #range(len(genome):\n",
    "        c,m = base2mask(g)\n",
    "        code = code + c\n",
    "        mask = mask + m\n",
    "        \n",
    "    #print(code,mask)\n",
    "    for i in range(0,len(code),64):\n",
    "        codeint.append(int(code[i:i+64],2))\n",
    "        maskint.append(int(mask[i:i+64],2))\n",
    "    return codeint, maskint\n",
    "\n",
    "genc,genm = genome2mask(genome)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Read the reads of the genome to compare"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read the reads on a FastQ file\n",
    "def readfile(fname):\n",
    "    readsg = ''\n",
    "    with open(fname) as fp:\n",
    "        line = fp.readline()\n",
    "        while line:\n",
    "            line = fp.readline()\n",
    "            readsg = readsg + line\n",
    "            line = fp.readline()\n",
    "            line = fp.readline()\n",
    "            line = fp.readline()\n",
    "    return readsg\n",
    "\n",
    "# Encode the genome string of the reads\n",
    "def get_reads(readsg):\n",
    "    reads = ''\n",
    "    for r in readsg.split('\\n'):\n",
    "        #gr = genome2bits(r)\n",
    "        reads = reads + r + '\\n'\n",
    "    reads = reads.split('\\n')[:-2]\n",
    "    return reads\n",
    "\n",
    "fname = 'bp100/bp100'\n",
    "readsg = readfile('reads/' + fname + \".bfast.fastq\")\n",
    "readsg = get_reads(readsg)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute for the hamming distance between two ternary strings\n",
    "def hamming_distance(str1, str2):\n",
    "    n = len(str1)\n",
    "    m = len(str2)\n",
    "    #print(n,m)\n",
    "    hamm = 0\n",
    "    for i in range(n):\n",
    "        if not (str1[i] == str2[i] or str1[i] == '*' or str2[i] == '*'):\n",
    "            hamm = hamm + 1\n",
    "    return hamm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Mimics the TCAM hardware\n",
    "class TCAM_Arrays:\n",
    "    \n",
    "    def __init__(self,nrow, ncolumn, error_threshold=4):\n",
    "        self.nrow = nrow\n",
    "        self.ncolumn = ncolumn\n",
    "        self.narray = 0\n",
    "        self.array = []\n",
    "        self.array_code = []\n",
    "        self.array_mask = []\n",
    "        self.tag = [] \n",
    "        self.error_threshold = error_threshold\n",
    "    \n",
    "    # Write on multiple TCAM arrays\n",
    "    # TODO: Create a function that writes on a specific row. (not necessary for genome sequencing)\n",
    "    def write_masked(self,genc,genm):\n",
    "        m = len(genc) % self.ncolumn \n",
    "        genc = genc + [0]*(self.ncolumn -m)\n",
    "        genm = genm + [9223372036854775807]*(self.ncolumn -m)\n",
    "        self.array_code = [genc[i:i+self.ncolumn] for i in range(0, len(genc), self.ncolumn)]\n",
    "        self.array_mask = [genm[i:i+self.ncolumn] for i in range(0, len(genm), self.ncolumn)]\n",
    "        self.narray = len(self.array_code)\n",
    "        tag = [[0] * self.nrow for i in range(self.narray)]\n",
    "    \n",
    "    \n",
    "    def write(self,bits):\n",
    "        nrc = self.nrow*self.ncolumn\n",
    "        m = len(bits) % nrc \n",
    "        bits = bits + \"0\"*(nrc -m)\n",
    "        bitsrow = [bits[i:i+self.nrow] for i in range(0, len(bits), self.nrow)]\n",
    "        self.array = [bitsrow[i:i+self.ncolumn] for i in range(0, len(bitsrow), self.ncolumn)]\n",
    "        self.narray = len(self.array)\n",
    "        self.tag = [[0] * nrow for i in range(self.narray)]\n",
    "    \n",
    "    def reset_tag(self):\n",
    "        self.tag = [[0] * nrow for i in range(self.narray)]\n",
    "        \n",
    "    def get_tag(self):\n",
    "        return self.tag\n",
    "    \n",
    "    # Implement a TCAM search of the query in all the arrays\n",
    "    # Returns:\n",
    "    #    pos  = positions (array, column, hamming distance=0) of exact matches\n",
    "    #    pos2 = positions (array, column, hamming distance) of partial matches \n",
    "    #              with hamming distance less than the error_threshold (default = 4)\n",
    "    def search(self,query):\n",
    "        na = 0\n",
    "        pos = []\n",
    "        pos2 = []\n",
    "        for a in self.array:\n",
    "            nc = 0\n",
    "            for ac in a:\n",
    "                if ac != '*'*64:\n",
    "                    e = hamming_distance(ac,query)\n",
    "                    if e == 0:\n",
    "                        pos.append([na, nc, e])\n",
    "                        self.tag[na][nc] = True\n",
    "                    elif e <= self.error_threshold:\n",
    "                        pos2.append([na, nc, e])\n",
    "                        self.tag[na][nc] = 1\n",
    "                nc = nc + 1\n",
    "            na = na + 1\n",
    "        return pos, pos2\n",
    "\n",
    "    def search_bits(self,code,mask):\n",
    "        na = 0\n",
    "        pos = []\n",
    "        pos2 = []\n",
    "        #print(self.narray,self.ncolumn)\n",
    "        for i in range(0,self.narray):\n",
    "            nc = 0\n",
    "            for j in range(0,self.ncolumn):\n",
    "                if self.array_mask[i][j] != 9223372036854775807:\n",
    "                    e = self.hamming_distance_mask(self.array_code[i][j],self.array_mask[i][j],code,mask)\n",
    "                    if e == 0:\n",
    "                        pos.append([na, nc, e])\n",
    "                        self.tag[na][nc] = True\n",
    "                    elif e <= self.error_threshold:\n",
    "                        pos2.append([na, nc, e])\n",
    "                        self.tag[na][nc] = 1\n",
    "                nc = nc + 1\n",
    "            na = na + 1\n",
    "        return pos, pos2\n",
    "    \n",
    "    # Find the array with the most tag\n",
    "    def best_array(self):\n",
    "        bestarray = [sum(array) for array in tcam.get_tag()]\n",
    "        max_bestarray = np.argmax(bestarray)\n",
    "        return max_bestarray\n",
    "        \n",
    "    # Find the rows with matches in the best_array\n",
    "    def best_match(self):\n",
    "        res = [rows for rows, val in enumerate(self.tag[self.best_array()]) if val > 0] \n",
    "        best_match = [indexi[j], -1, int(max_bestarray), [res]]\n",
    "        return best_match\n",
    "    \n",
    "\n",
    "    def hamming_distance_mask(self,ac,am,bc,bm):\n",
    "        hamm = bin(ac ^ bc & ~am & ~bm).count('1')\n",
    "        #print(ac,bc,am,bm,hamm)\n",
    "        return hamm\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(0, ' of ', 10)\n",
      "[[4147, [45, 46, 47, 48, 49]], [4147, [45, 46, 47, 48, 49]]]\n",
      "(1, ' of ', 10)\n",
      "[[2635, [50, 51, 52, 53, 54]], [2635, [50, 51, 52, 53, 54]]]\n",
      "(2, ' of ', 10)\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-10-67634e62d263>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     78\u001b[0m \u001b[0mtcam\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwrite_masked\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgenc\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mgenm\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     79\u001b[0m \u001b[0mindexi\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mrange\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;36m10\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 80\u001b[0;31m \u001b[0mpos_100\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpos2_100\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mbestmatch\u001b[0m\u001b[0;34m=\u001b[0m \u001b[0mfindreads\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtcam\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mreads\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;36m20\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindexi\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'results/'\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mfname\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;34m'.pos'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     81\u001b[0m \u001b[0;31m#pos_100, pos2_100, bestmatch = findreads(tcam, reads[0:20], indexi, 'results/' + fname + '.posr', reverse = True)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     82\u001b[0m \u001b[0;31m#print(tcam.array_code[0])\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-10-67634e62d263>\u001b[0m in \u001b[0;36mfindreads\u001b[0;34m(tcam, reads, indexi, pos_fname)\u001b[0m\n\u001b[1;32m     36\u001b[0m                     \u001b[0mqcode\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mqcode\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     37\u001b[0m                     \u001b[0mqmask\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mqmask\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 38\u001b[0;31m                 \u001b[0mpos\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpos2\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtcam\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msearch_bits\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mqcode\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mqmask\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     39\u001b[0m                 \u001b[0;31m#print(len(pos),len(pos2))\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     40\u001b[0m                 \u001b[0;31m#print(pos2)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-6-236cf08882d9>\u001b[0m in \u001b[0;36msearch_bits\u001b[0;34m(self, code, mask)\u001b[0m\n\u001b[1;32m     72\u001b[0m             \u001b[0;32mfor\u001b[0m \u001b[0mj\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mrange\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mncolumn\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     73\u001b[0m                 \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0marray_mask\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mj\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0;36m9223372036854775807\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 74\u001b[0;31m                     \u001b[0me\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mhamming_distance_mask\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0marray_code\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mj\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0marray_mask\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mj\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mcode\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mmask\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     75\u001b[0m                     \u001b[0;32mif\u001b[0m \u001b[0me\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     76\u001b[0m                         \u001b[0mpos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mna\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-6-236cf08882d9>\u001b[0m in \u001b[0;36mhamming_distance_mask\u001b[0;34m(self, ac, am, bc, bm)\u001b[0m\n\u001b[1;32m     97\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     98\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mhamming_distance_mask\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mac\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mam\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mbc\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mbm\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 99\u001b[0;31m         \u001b[0mhamm\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mbin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mac\u001b[0m \u001b[0;34m^\u001b[0m \u001b[0mbc\u001b[0m \u001b[0;34m&\u001b[0m \u001b[0;34m~\u001b[0m\u001b[0mam\u001b[0m \u001b[0;34m&\u001b[0m \u001b[0;34m~\u001b[0m\u001b[0mbm\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcount\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'1'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    100\u001b[0m         \u001b[0;31m#print(ac,bc,am,bm,hamm)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    101\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mhamm\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "# Save positional information\n",
    "def save_pos(pos, pos_fname, prefix=''):\n",
    "    with open(pos_fname, \"a\") as f:\n",
    "        for s in pos:\n",
    "            f.write(prefix + str(s[0]) + ' ' + str(s[1]) + ' ' + str(s[2]) +\"\\n\")\n",
    "\n",
    "def findreads(tcam, reads, indexi, pos_fname):\n",
    "    \n",
    "    pos = ''\n",
    "    pos2 = ''\n",
    "    ismatch = [False] * len(reads)    \n",
    "    best_matches = []\n",
    "\n",
    "    #for j in  readsi: \n",
    "    f = open(pos_fname + '.best','w')\n",
    "\n",
    "    for j in  range(len(indexi)):\n",
    "        print(j, \" of \", len(indexi))\n",
    "        tcam.reset_tag()\n",
    "        idx = indexi[j]\n",
    "        query = reads[idx]\n",
    "        best_matches = []\n",
    "        best_match = [0] * 2\n",
    "            \n",
    "        for reverse in [False, True]:\n",
    "            matchflag = False\n",
    "            N = len(query) - 16\n",
    "            for i in range(0,N,1):\n",
    "                q = query[i:i+16]\n",
    "                #print(i, \" of \", N, \" shifts\", q)\n",
    "                qcode, qmask = genome2mask(q)\n",
    "                if reverse == True:\n",
    "                    qcode = int('{:064b}'.format(qcode[0])[::-1], 2)\n",
    "                    qmask = int('{:064b}'.format(qmask[0])[::-1], 2)\n",
    "                else:\n",
    "                    qcode = qcode[0]\n",
    "                    qmask = qmask[0]\n",
    "                pos, pos2 = tcam.search_bits(qcode,qmask)\n",
    "                #print(len(pos),len(pos2))\n",
    "                #print(pos2)\n",
    "                #save_pos(pos,pos_fname, 'Match ' + str(indexi[j]) + ' ' + str(i) + ' ')\n",
    "                #save_pos(pos2,pos_fname + '.par', 'Partial Match ' + str(indexi[j]) + ' ' + str(i) + ' ')\n",
    "                #if len(pos) > 0:\n",
    "                #    best_matches.append([indexi[j], i, -1, pos]) \n",
    "            ismatch[j] = matchflag\n",
    "\n",
    "            res = [rows for rows, val in enumerate(tcam.get_tag()[tcam.best_array()]) if val > 0] \n",
    "            #best_matches.append([indexi[j], -1, int(tcam.best_array()), [res]])\n",
    "            \n",
    "            if reverse == False:\n",
    "                best_match[0] = [int(tcam.best_array()), res]\n",
    "            else: \n",
    "                best_match[1] = [int(tcam.best_array()), res]\n",
    "                \n",
    "\n",
    "        print(best_match)\n",
    "        \n",
    "        if len(best_match[0][1]) >= len(best_match[1][1]):\n",
    "            position = (int(best_match[0][0])) *nrow*ncol/4 + (int(best_match[0][1][0]))*nrow/4 \n",
    "        else:\n",
    "            position = (int(best_match[1][0])) *nrow*ncol/4 + (int(best_match[1][1][0]))*nrow/4 \n",
    "            \n",
    "        best_matches.append([idx, position])\n",
    "    \n",
    "        for s in best_matches:\n",
    "            f.write(str(s[0]) + ',' + str(s[1]))\n",
    "            f.write('\\n')\n",
    "        #clear_output()\n",
    "            \n",
    "    f.close()\n",
    "    return pos, pos2, best_matches\n",
    "\n",
    "\n",
    "fname = 'bp100/bp100'\n",
    "readsg = readfile('reads/' + fname + \".bfast.fastq\")\n",
    "reads = get_reads(readsg)\n",
    "tcam = TCAM_Arrays(nrow,ncol)\n",
    "tcam.write_masked(genc,genm)\n",
    "indexi = range(0,10)\n",
    "pos_100, pos2_100, bestmatch= findreads(tcam, reads[0:20], indexi, 'results/' + fname + '.pos')\n",
    "#pos_100, pos2_100, bestmatch = findreads(tcam, reads[0:20], indexi, 'results/' + fname + '.posr', reverse = True)\n",
    "#print(tcam.array_code[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "a = 1426534132306446472\n",
    "qcode = int('{:064b}'.format(a)[::-1], 2)\n",
    "print(qcode)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.15+"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
